common/src/main/scala/notifications/providers/WindowsNotificationProvider.scala

```diff
@@ -23,10 +25,23 @@ class WindowsNotificationProvider(wsClient: WSClient, connectionString: String,
 
   private val hubClient = new NotificationHubClient(notificationHub, wsClient)
 
-  override def register(registration: Registration): Future[\/[Error, RegistrationResponse]] =
-    hubClient.register(RawWindowsRegistration.fromMobileRegistration(registration)).map { hubResult =>
-      hubResult.map { _.toRegistrarResponse }
+  override def register(registration: Registration): Future[\/[Error, RegistrationResponse]] = {
+    def createNewRegistration = hubClient
+      .create(RawWindowsRegistration.fromMobileRegistration(registration))
+      .map(hubResultToRegistrationResponse)
+
+    def updateRegistration(regId: WNSRegistrationId) = hubClient
+      .update(regId, RawWindowsRegistration.fromMobileRegistration(registration))
+      .map(hubResultToRegistrationResponse)
+
+    val channelUri = registration.deviceId
+    hubClient.registrationsByChannelUri(channelUri).flatMap {
+      case \/-(Nil) => createNewRegistration
+      case \/-(existing :: Nil) => updateRegistration(existing.registration)
+      case \/-(List(_, _)) => Future { TooManyRegistrationsForChannel(channelUri).left }
```

Comment:
You could replace List(_, _) with _
Or with first :: second :: tail
Instead of wrapping TooManyRegistrations in a Future, use Future.successful(...)
same for e.left